github/branches/retrieve_github_data.py

```diff
@@ -0,0 +1,610 @@
+            "app or personal token at "
+            "https://github.com/settings/tokens"
+        )
+    if not args.prod and (len(args.repo) == 0):
+        ap.error("Must supply 'repo' or set '--prod'")
+    return args
+
+
+# due to pytest parametrization, we'll call this many times for each
+# repo consecutively, so a large cache is not needed.
+@lru_cache(maxsize=2)
+def get_repo_branch_protections(endpoint, repo: str) -> RepoBranchProtections:
+    if repo.endswith(EXTENSION_TO_STRIP):
+        repo = repo[: -len(EXTENSION_TO_STRIP)]
+    data = get_nested_branch_data(endpoint, repo)
+    return data
+
+
+# helper classes for graph errors
+def _compact_fmt(d):
+    s = []
+    for k, v in d.items():
+        if isinstance(v, dict):
+            v = _compact_fmt(v)
+        elif isinstance(v, (list, tuple)):
+            lst = []
+            for e in v:
+                if isinstance(e, dict):
+                    lst.append(_compact_fmt(e))
+                else:
+                    lst.append(repr(e))
+            s.append("{}=[{}]".format(k, ", ".join(lst)))
+            continue
+        s.append(f"{k}={v!r}")
+    return "(" + ", ".join(s) + ")"
+
+
+def _report_download_errors(errors):
+    """error handling for graphql comms."""
+    logger.error("Document contain %d errors", len(errors))
+    for i, e in enumerate(errors):
+        msg = e.pop("message")
+        extra = ""
+        if e:
+            extra = " %s" % _compact_fmt(e)
+        logger.error("Error #%d: %s%s", i + 1, msg, extra)
+
+
+def get_connection(base_url: str, token: str) -> Any:
+    endpoint = HTTPEndpoint(base_url, {"Authorization": "bearer " + token,})
+    endpoint.report_download_errors = _report_download_errors
+
+    # determine which date we're collecting for
+    global _collection_date
+    assert _collection_date == "1970-01-01"
+    _collection_date = date.today().isoformat()
+    return endpoint
+
+
+def _in_offline_mode() -> bool:
+    is_offline = False
+    try:
+        # if we're running under pytest, we need to fetch the value from
+        # the current configuration
+        import conftest
+
+        is_offline = conftest.get_client("github_client").is_offline()
+        if not is_offline:
+            # check for a valid GH_TOKEN here so we fail during test collection
+            # TODO: make sure this works in all scenarios
+            os.environ["GH_TOKEN"]
+    except ImportError:
+        pass
+
+    return is_offline
+
+
+def repos_to_check() -> List[BranchOfInterest]:
+    # just shell out for now
+    # TODO: fix ickiness
+    #   While there is no network operation done here, we don't want to go
+    #   poking around the file system if we're in "--offline" mode
+    #   (e.g. doctest mode)
+    if _in_offline_mode():
+        return []
+
+    # real work
+    # DEV_HACK: find better way to insert dev default
+    path_to_metadata = os.environ.setdefault(
+        "PATH_TO_METADATA", "~/repos/foxsec/master/services/metadata"
```

Comment:
Should probably provide this path as an option?

Reply:
And, I think an option to the groovy pipeline.